src/inspector/processing/data-processing.service.ts

```diff
@@ -346,34 +334,50 @@ export class DataProcessingService implements OnModuleInit {
     });
   }
 
+  savedCanonSlotsAttProperties = {};
+
+  protected async getCanonSlotRoot(slot: bigint) {
+    const cached = this.savedCanonSlotsAttProperties[String(slot)];
+    if (cached) return cached;
+    const root = (await this.clClient.getBeaconBlockHeaderOrPreviousIfMissed(slot))[0].root;
+    this.savedCanonSlotsAttProperties[String(slot)] = root;
+    return root;
+  }
+
   protected async prepAttestationsToWrite(attDutyResult: CheckAttestersDutyResult) {
-    for (const a of attDutyResult.attestersDutyInfo) {
-      a.attested = false;
-      a.in_block = undefined;
-      for (const [block, blockAttestations] of Object.entries(attDutyResult.blocksAttestations)) {
-        if (BigInt(a.slot) >= BigInt(block)) continue; // Attestation cannot be included in the previous or current block
-        const committeeAttestationInfo: any = blockAttestations.find(
-          (att: any) => att.slot == a.slot && att.committee_index == a.committee_index,
+    this.savedCanonSlotsAttProperties = {};
+    const blocksAttestation = Object.entries(attDutyResult.blocksAttestations).sort(
+      (b1, b2) => parseInt(b1[0]) - parseInt(b2[0]), // Sort array by block number
+    );
+    for (const duty of attDutyResult.attestersDutyInfo) {
+      duty.attested = false;
+      duty.valid_head = undefined;
+      duty.valid_target = undefined;
+      duty.valid_source = undefined;
+      duty.in_block = undefined;
+      for (const [block, blockAttestations] of blocksAttestation.filter(
+        ([b]) => BigInt(b) > BigInt(duty.slot), // Attestation cannot be included in the previous or current block
+      )) {
+        const committeeAttestations: SlotAttestation[] = blockAttestations.filter(
+          (att: any) => att.slot == duty.slot && att.committee_index == duty.committee_index,
         );
-        if (committeeAttestationInfo) {
-          a.attested = committeeAttestationInfo.bits[parseInt(a.validator_committee_index)];
-          a.in_block = block;
-        }
-        if (a.attested) {
-          // We found the nearest block includes validator attestation
+        if (!committeeAttestations) continue;
+        for (const ca of committeeAttestations) {
+          duty.attested = ca.bits[parseInt(duty.validator_committee_index)];
+          duty.in_block = block;
+          if (!duty.attested) continue; // continue to find attestation with validator attestation
+          // and if validator attests block - calculate inclusion delay and check properties (head, target, source)
+          // calculate inclusion delay
           const missedSlotsOffset = attDutyResult.allMissedSlots.filter(
```

Comment:
Empty slots?

Reply:
Expand the question pls

Reply:
What `const missedSlotsOffset` actually means? Empty slots before the missed attestation?

Reply:
If so may be it worth to update the name.

Reply:
It means count of missed slots between slot to attestation and block where this attestation is included. According to spec, we must exclude missed slots when calculating inclusion delay

Reply:
I understand offset as something continuous, but according to the description it is just a count, so I suggest `missedSlotsCount` variable name.

Reply:
done